File: src/main/java/com/williambl/haema/vampire/VampireComponent.java
package com.williambl.haema.vampire;

import dev.onyxstudios.cca.api.v3.component.Component;
import dev.onyxstudios.cca.api.v3.component.ComponentKey;
import dev.onyxstudios.cca.api.v3.component.ComponentRegistry;

import static com.williambl.haema.Haema.id;

/**
 * Handles basic vampire state for an entity.
 */
public interface VampireComponent extends Component {
    /**
     * Get whether the entity is a vampire.
     * @return whether the entity is a vampire
     */
    boolean isVampire();

    /**
     * Set whether the entity is a vampire.
     * @param vampire whether the entity is a vampire
     */
    void setVampire(boolean vampire);

    /**
     * Get the entity's blood level.
     * @return the entity's blood level
     */
    double getBlood();

    /**
     * Set the entity's blood level.
     * @param blood the entity's blood level
     */
    void setBlood(double blood);

    ComponentKey<VampireComponent> KEY = ComponentRegistry.getOrCreate(id("vampire"), VampireComponent.class);
}

File: src/main/kotlin/com/williambl/haema/component/VampireComponent.kt
package com.williambl.haema.component

import com.williambl.haema.ability.VampireAbility
import com.williambl.haema.id
import dev.onyxstudios.cca.api.v3.component.ComponentKey
import dev.onyxstudios.cca.api.v3.component.ComponentRegistryV3
import dev.onyxstudios.cca.api.v3.component.ComponentV3
import dev.onyxstudios.cca.api.v3.component.tick.ServerTickingComponent
import net.minecraft.entity.LivingEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.Identifier

interface VampireComponent : ComponentV3, ServerTickingComponent {

    var isVampire: Boolean
    var isPermanentVampire: Boolean
    var isKilled: Boolean

    var absoluteBlood: Double
    val blood: Double
    val lastFed: Long

    var abilities: MutableMap<VampireAbility, Int>
    var ritualsUsed: MutableSet<Identifier>

    fun removeBlood(blood: Double)
    fun addBlood(blood: Double)
    fun feed(entity: LivingEntity): ActionResult

    companion object {
        val entityKey: ComponentKey<VampireComponent> = ComponentRegistryV3.INSTANCE.getOrCreate(id("vampire"), VampireComponent::class.java)
    }
}
